presto-main/src/main/java/com/facebook/presto/operator/OperatorContext.java

```diff
@@ -409,24 +413,41 @@ public synchronized long requestMemoryRevoking()
     {
         boolean alreadyRequested = isMemoryRevokingRequested();
         if (!alreadyRequested && revocableMemoryReservation > 0) {
-            memoryRevokingRequestedFuture.set(null);
+            memoryRevokingRequested.set(true);
+            runListener();
             return revocableMemoryReservation;
         }
         return 0;
     }
 
     public synchronized void resetMemoryRevokingRequested()
     {
-        SettableFuture<?> currentFuture = memoryRevokingRequestedFuture;
-        if (!currentFuture.isDone()) {
-            return;
+        if (memoryRevokingRequested.compareAndSet(true, false)) {
+            memoryRevokingRequestedListener.set(null);
         }
-        memoryRevokingRequestedFuture = SettableFuture.create();
     }
 
-    public synchronized SettableFuture<?> getMemoryRevokingRequestedFuture()
+    public synchronized void setMemoryRevokingRequestedListener(Runnable listener)
     {
-        return memoryRevokingRequestedFuture;
+        verify(memoryRevokingRequestedListener.get() == null, "A listener has already been set");
```

Comment:
I don't follow how do we prevent this `verify` to trigger.
Don't we call `setMemoryRevokingRequestedListener `  always (almost) in `wakeUpOnRevokeRequest` for `OperatorContext`s where listeners are already registered?